File: python/36_Valid_Sudoku.py
import collections

class Valid_Sudoku:
  def isValidSudoku(self, board: list[list[str]]) -> bool:
    rows = collections.defaultdict(set)
    cols = collections.defaultdict(set)
    squares = collections.defaultdict(set)
    for r in range(9):
      for c in range(9):
        ch = board[r][c]
        print(ch)
        if ch == '.':
          continue
        if (ch in rows[r] or ch in cols[c] or ch in squares[(r//3,c//3)]):
          return False
        rows[r].add(ch)
        cols[c].add(ch)
        squares[(r//3,c//3)].add(ch)
    return True

File: c++/36_Valid_Sudoku.cpp
#include <set>
#include <vector>

using namespace std;

class Valid_Sudoku {
public:
  bool isValidSudoku(vector<vector<char>>& board) {
    set<int> rows[10], cols[10];
    set<int> squares[3][3];

    for (int r = 0; r <= 8; r++) {
      for (int c = 0; c <= 8; c++) {
        char ch = board[r][c];
        if (ch == '.'){
          continue;
        }
        if (rows[r].count(ch) || cols[c].count(ch) || squares[int(r/3)][int(c/3)].count(ch)) {
          return false;
        }
        rows[r].insert(ch);
        cols[c].insert(ch);
        squares[int(r/3)][int(c/3)].insert(ch);
      }
    }
    return true;
  }
};